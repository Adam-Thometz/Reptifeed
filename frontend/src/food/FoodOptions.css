.FoodOptions {
  border: 2px solid #D5CE3B;
  width: 50vw;
  height: 60%;
  border-radius: 20px;
  margin: 0 auto;
  padding: 1%;
  background-color: #3E3E07;
  color: #D5CE3B;
  box-shadow: 10px 5px 5px rgb(4, 66, 4);
  padding: 20px;
  margin: 20px auto;
}

.FoodOptions-group {
  width: 70%;
  margin: 0 auto;
  padding: 10px;
  display: flex;
}

.FoodOptions-label {
  width: 35%;
  margin: 0 1%;
  padding: auto 1%;
}

.FoodOptions-input {
  width: 65%;
  padding: 1%;
  border-radius: 5px;
  border: none;
  margin: 5px;
}

input[type="checkbox"] {
  width: 20px;
  height: 20px;
}