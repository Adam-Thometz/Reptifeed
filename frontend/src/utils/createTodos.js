
/** createTodos(reptiles, pantry)
 * 
 * Takes a user's reptiles and pantry and creates a list of things they need to do to unlock the app's full features
 * 
 * Returns an two arrays of Link components, one of important things and one of nice-to-haves
 */

function createTodos(reptiles, pantry) {
  /** Essential checks:
   * - reptileCheck: Checks if a user has reptiles
   * - vegetableOftenCheck: Checks if a user has a vegetable staple in their pantry
   * - proteinOftenCheck: Checks if a user has a protein staple in their pantry
   * - treatModeratelyCheck: Checks if a user has a treat staple in their pantry
   * - supplementCheck: Checks if a user has at least two different supplements
   */
  const essentialTodos = [];

  const reptileCheck = !!reptiles.length;
  const vegetableOftenCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'often');
  const proteinOftenCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'often');
  const treatModeratelyCheck = pantry.some(food => food.isTreat && food.frequency === 'moderately');
  const supplementCheck = (pantry.filter(food => food.type === 'supplement')).length >= 2;

  if (!reptileCheck) essentialTodos.push('Add a reptile');
  if (!vegetableOftenCheck) essentialTodos.push('Add vegetables eaten often');
  if (!proteinOftenCheck) essentialTodos.push('Add protein eaten often');
  if (!treatModeratelyCheck) essentialTodos.push('Add a treat eaten moderately');
  if (!supplementCheck) essentialTodos.push('Add at least two supplements');

  /** Nice-to-have checks:
   * - vegetableModeratelyCheck: Checks if a user has a vegetable that's eaten moderately in their pantry
   * - vegetableOccasionallyCheck: Checks if a user has a vegetable that's eaten occasionally in their pantry
   * - proteinModeratelyCheck: Checks if a user has a protein that's eaten moderately in their pantry
   * - proteinOccasionallyCheck: Checks if a user has a protein that's eaten occasionally in their pantry
   * - treatOccasionallyCheck: Checks if a user has a treat that's eaten rarely in their pantry
   */
  const niceToHaveTodos = [];

  const vegetableModeratelyCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'moderately');
  const vegetableOccasionallyCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'occasionally' && !food.isTreat);
  const proteinModeratelyCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'moderately');
  const proteinOccasionallyCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'occasionally' && !food.isTreat);
  const treatOccasionallyCheck = pantry.some(food => food.isTreat && food.frequency === 'occasionally');

  if (!vegetableModeratelyCheck) niceToHaveTodos.push('Add a vegetable eaten moderately');
  if (!vegetableOccasionallyCheck) niceToHaveTodos.push('Add a vegetable eaten occasionally');
  if (!proteinModeratelyCheck) niceToHaveTodos.push('Add a protein eaten moderately');
  if (!proteinOccasionallyCheck) niceToHaveTodos.push('Add a protein eaten occasionally');
  if (!treatOccasionallyCheck) niceToHaveTodos.push('Add a treat eaten occasionally');

  return {essentialTodos, niceToHaveTodos};
};

export default createTodos;