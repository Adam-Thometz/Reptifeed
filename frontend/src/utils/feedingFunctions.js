/** Feeding functions
 * 
 * Used to get foods from pantry that match frequencies and food type.
*/

function getFoodFromPantry(pantry, freq, type) {
  const availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === freq && food.type === type)));
  if (!availableFood.length) return `Get some more ${type}`
  const selectedIdx = Math.floor(Math.random() * availableFood.length);

  return availableFood[selectedIdx];
};

function getTreatFromPantry(pantry, freq) {
  const availableTreats = pantry.filter(food => food.isTreat && food.frequency === freq);
  if (!availableTreats.length) return "You don't have any treats";
  const selectedIdx = Math.floor(Math.random() * availableTreats.length); 
  
  return availableTreats[selectedIdx];
};

export { getFoodFromPantry, getTreatFromPantry };