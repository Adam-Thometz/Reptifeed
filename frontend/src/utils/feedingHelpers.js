/** Feeding functions
 * 
 * Used to get foods from pantry that match frequencies and food type.
 * 
 * If no foods are available for the frequency, returns foods eaten most often
*/

function getFoodFromPantry(pantry, freq, type) {
  let availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === freq && food.type === type)));
  if (!availableFood.length) {
    availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === 'often' && food.type === type)));
  };
  const selectedIdx = Math.floor(Math.random() * availableFood.length);

  return availableFood[selectedIdx];
};

function getTreatFromPantry(pantry, freq) {
  let availableTreats = pantry.filter(food => food.isTreat && food.frequency === freq);
  if (!availableTreats.length) {
    availableTreats = pantry.filter(food => food.isTreat);
  };
  const selectedIdx = Math.floor(Math.random() * availableTreats.length); 
  
  return availableTreats[selectedIdx];
};

function getNextSupplement(pantry, lastSupplement) {
  const nextSupplement = lastSupplement === 'calcium' ?
    pantry.find(food => food.name === 'multivitamin') :
    pantry.find(food => food.name === 'calcium');
  
  return nextSupplement;
}

export { getFoodFromPantry, getTreatFromPantry, getNextSupplement };