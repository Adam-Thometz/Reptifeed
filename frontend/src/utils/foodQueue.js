// foodQueue: Feed reptile based on frequency
const foodQueue = ["often", "often", "moderately", "often", "often", "moderately", "often", "often", "moderately", "occasionally"];

// treatQueue: Give reptile treats based on frequency
const treatQueue = ["moderately", "moderately", "occasionally"];

export { foodQueue, treatQueue };