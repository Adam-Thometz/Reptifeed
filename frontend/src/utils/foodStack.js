// foodStack: Feed reptile based on frequency
const foodStack = ["occasionally", "moderately", "often", "often", "moderately", "often", "often", "moderately", "often", "often"];

// treatStack: Give reptile treats based on frequency
const treatStack = ["occasionally", "moderately", "moderately"];

export { foodStack, treatStack };