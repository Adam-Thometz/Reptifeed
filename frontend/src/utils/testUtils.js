import React from "react";
import UserContext from "./UserContext";

const demoUser = {
  id: 1,
  username: 'rickSanchez',
  password: 'ilikeMorty',
  email: 'rickman@qmail.com',
  isAdmin: false
};

const demoReptile = [
  {
    id: 1,
    name: 'two crows',
    species: 'crows',
    subspecies: 'two of them',
    birthday: '2015-10-29',
    imgUrl: '',
    ownerId: 1
  }
];

const demoPantry = [
  {
    name: 'plumbus',
    type: 'everything',
    frequency: 'always',
    image: '',
    isTreat: true,
    tips: "Everyone needs a plumbus!"
  }
];

const UserProvider = ({ children, currUser = demoUser, pantry = demoPantry, reptiles = demoReptile }) => (
  <UserContext.Provider value={{currUser, pantry, reptiles}}>
    {children}
  </UserContext.Provider>
);

export { UserProvider };