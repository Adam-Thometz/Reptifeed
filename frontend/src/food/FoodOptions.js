import React from "react";
import './FoodOptions.css'

const FoodOptions = ({species, setSpecies, searchTerm, setSearchTerm, foodFilter, setFoodFilter, treatFilter, setTreatFilter}) => {
  const handleChangeSpecies = e => {
    setSpecies(e.target.value)
  };

  const handleChangeSearch = e => {
    setSearchTerm(e.target.value);
  }
  
  const handleChangeType = e => {
    setFoodFilter(e.target.value);
  };

  const handleChangeTreat = () => {
    setTreatFilter(treat => !treat);
  };

  return (
    <div className="FoodOptions">
      <h4>Use these options to filter foods</h4>
      <div className="FoodOptions-group">
        <label className="FoodOptions-label" htmlFor="species">Species</label>
        <select
          className="FoodOptions-input"
          id="species"
          name="species"
          value={species}
          onChange={handleChangeSpecies}
        >
          <option value="blue-tongue-skink">Blue Tongue Skink</option>
        </select>
      </div>
      <div className="FoodOptions-group">
        <label className="FoodOptions-label" htmlFor="search">Search foods</label>
        <input
        type="text"
          className="FoodOptions-input"
          id="search"
          name="search"
          value={searchTerm}
          onChange={handleChangeSearch}
        />
      </div>
      <div className="FoodOptions-group">
        <label className="FoodOptions-label" htmlFor="frequency">Food group</label>
        <select
          className="FoodOptions-input"
          id="type"
          name="type"
          value={foodFilter}
          onChange={handleChangeType}
        >
          <option value="">--</option>
          <option value="protein">Protein</option>
          <option value="vegetable">Vegetables/Greens</option>
          <option value="fruit">Fruit</option>
          <option value="supplement">Supplements</option>
        </select>
      </div>
      <div className="FoodOptions-group">
        <label className="FoodOptions-label" htmlFor="treats">Treats</label>
        <input
          className="FoodOptions-checkbox"
          type="checkbox"
          id="treats"
          name="treats"
          value={treatFilter}
          onChange={handleChangeTreat}
        />
      </div>
    </div>
  )
};

export default FoodOptions;