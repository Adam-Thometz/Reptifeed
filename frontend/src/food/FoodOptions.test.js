import React from "react";
import { render } from '@testing-library/react'
import FoodOptions from "./FoodOptions";
import { MemoryRouter } from "react-router-dom";
import { UserProvider } from '../utils/testUtils'

test('it renders without crashing', () => {
  render(<MemoryRouter>
    <UserProvider>
      <FoodOptions />
    </UserProvider>
  </MemoryRouter>);
});

test('it matches snapshot', () => {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <FoodOptions />
      </UserProvider>
    </MemoryRouter>);
  expect(asFragment()).toMatchSnapshot();
})