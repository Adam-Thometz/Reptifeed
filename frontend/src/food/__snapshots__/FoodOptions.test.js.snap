// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="FoodOptions"
  >
    <p>
      Use these options to filter foods
    </p>
    <div
      class="FoodOptions-group"
    >
      <label
        class="FoodOptions-label"
        for="species"
      >
        Species
      </label>
      <select
        class="FoodOptions-input"
        id="species"
        name="species"
      >
        <option
          value="blue-tongue-skink"
        >
          Blue Tongue Skink
        </option>
      </select>
    </div>
    <div
      class="FoodOptions-group"
    >
      <label
        class="FoodOptions-label"
        for="search"
      >
        Search foods
      </label>
      <input
        class="FoodOptions-input"
        id="search"
        name="search"
        type="text"
        value=""
      />
    </div>
    <div
      class="FoodOptions-group"
    >
      <label
        class="FoodOptions-label"
        for="frequency"
      >
        Food group
      </label>
      <select
        class="FoodOptions-input"
        id="type"
        name="type"
      >
        <option
          value=""
        >
          --
        </option>
        <option
          value="protein"
        >
          Protein
        </option>
        <option
          value="vegetable"
        >
          Vegetables/Greens
        </option>
        <option
          value="fruit"
        >
          Fruit
        </option>
        <option
          value="supplement"
        >
          Supplements
        </option>
      </select>
    </div>
    <div
      class="FoodOptions-group"
    >
      <label
        class="FoodOptions-label"
        for="treats"
      >
        Treats
      </label>
      <input
        class="FoodOptions-checkbox"
        id="treats"
        name="treats"
        type="checkbox"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
